bpf/lib/egress_policies.h

```diff
@@ -28,6 +28,81 @@ struct egress_gw_policy_entry *lookup_ip4_egress_gw_policy(__be32 saddr, __be32
 	return map_lookup_elem(&EGRESS_POLICY_MAP, &key);
 }
 
+static __always_inline
+bool egress_gw_request_needs_redirect(struct iphdr *ip4, __u32 dst_id,
+				      __u32 *tunnel_endpoint)
+{
+	struct egress_gw_policy_entry *egress_gw_policy;
+	struct endpoint_info *gateway_node_ep;
+
+	/* If the packet is destined to an entity inside the cluster,
+	 * either EP or node, it should not be forwarded to an egress
+	 * gateway since only traffic leaving the cluster is supposed to
+	 * be masqueraded with an egress IP.
+	 */
+	if (identity_is_cluster(dst_id))
+		return false;
+
+	egress_gw_policy = lookup_ip4_egress_gw_policy(ip4->saddr, ip4->daddr);
+	if (!egress_gw_policy)
+		return false;
+
+	/* If the gateway node is the local node, then just let the
+	 * packet go through, as it will be SNATed later on by
+	 * handle_nat_fwd().
+	 */
+	gateway_node_ep = __lookup_ip4_endpoint(egress_gw_policy->gateway_ip);
+	if (gateway_node_ep && (gateway_node_ep->flags & ENDPOINT_F_HOST))
+		return false;
+
+	*tunnel_endpoint = egress_gw_policy->gateway_ip;
+	return true;
+}
+
+static __always_inline
+bool egress_gw_snat_needed(struct iphdr *ip4, struct remote_endpoint_info *remote_ep,
+			   __be32 *snat_addr)
+{
+	struct egress_gw_policy_entry *egress_gw_policy;
+
+	/* If the packet is destined to an entity inside the cluster, either EP
+	 * or node, skip SNAT since only traffic leaving the cluster is supposed
+	 * to be masqueraded with an egress IP.
+	 */
+	if (remote_ep &&
+	    identity_is_cluster(remote_ep->sec_label))
+		return false;
+
+	egress_gw_policy = lookup_ip4_egress_gw_policy(ip4->saddr, ip4->daddr);
+	if (!egress_gw_policy)
+		return false;
+
+	*snat_addr = egress_gw_policy->egress_ip;
+	return true;
+}
+
+static __always_inline
+bool egress_gw_reply_needs_redirect(struct iphdr *ip4, __u32 *tunnel_endpoint,
+				    __u32 *dst_id)
+{
+	struct egress_gw_policy_entry *egress_policy;
+
+	/* Find a matching policy by looking up the reverse address tuple: */
+	egress_policy = lookup_ip4_egress_gw_policy(ip4->daddr, ip4->saddr);
+	if (egress_policy) {
+		struct remote_endpoint_info *info;
+
+		info = ipcache_lookup4(&IPCACHE_MAP, ip4->daddr, V4_CACHE_KEY_LEN);
+		if (info && info->tunnel_endpoint != 0) {
+			*tunnel_endpoint = info->tunnel_endpoint;
+			*dst_id = info->sec_label;
+			return true;
+		}
+	}
+
+	return false;
```

Comment:
Nit since we are on this, feel free to ignore it:

```suggestion
	struct egress_gw_policy_entry *egress_policy;
	struct remote_endpoint_info *info;

	/* Find a matching policy by looking up the reverse address tuple: */
	egress_policy = lookup_ip4_egress_gw_policy(ip4->daddr, ip4->saddr);
	if (!egress_policy)
		return false;
		
	info = ipcache_lookup4(&IPCACHE_MAP, ip4->daddr, V4_CACHE_KEY_LEN);
	if (!info || !info->tunnel_endpoint)
		return false;
		
	*tunnel_endpoint = info->tunnel_endpoint;
	*dst_id = info->sec_label;
	
	return true;
```

(imo more readable and consistent with the style of the rest of the logic)

Reply:
Ack, that's my preferred style too :+1:.

Reply:
Done.